hy/macros.py

```diff
@@ -48,88 +35,144 @@ def _(fn):
+
         if not PY3:
             _name = _name.encode('UTF-8')
+
         fn.__name__ = _name
-        module_name = fn.__module__
+
+        module = inspect.getmodule(fn)
+
+        module_name = module.__name__
         if module_name.startswith("hy.core"):
             module_name = None
-        _hy_tag[module_name][mangle(name)] = fn
+
+        module_tags = module.__dict__.setdefault('__tags__', {})
+        module_tags[mangle(name)] = fn
 
         return fn
     return _
 
 
 def require(source_module, target_module, assignments, prefix=""):
-    """Load macros from `source_module` in the namespace of
-    `target_module`. `assignments` maps old names to new names, or
-    should be the string "ALL". If `prefix` is nonempty, it is
-    prepended to the name of each imported macro. (This means you get
-    macros named things like "mymacromodule.mymacro", which looks like
-    an attribute of a module, although it's actually just a symbol
-    with a period in its name.)
+    """Load macros from one module into the namespace of another.
 
     This function is called from the `require` special form in the compiler.
 
-    """
+    Parameters
+    ----------
+    source_module: str or types.ModuleType
+        The module from which macros are to be imported.
 
-    seen_names = set()
-    if prefix:
-        prefix += "."
-    if assignments != "ALL":
-        assignments = {mangle(str_type(k)): v for k, v in assignments}
+    target_module: str, types.ModuleType or None
+        The module into which the macros will be loaded.  If `None`, then
+        the caller's namespace.
+        The latter is useful during evaluation of generated AST/bytecode.
 
-    for d in _hy_macros, _hy_tag:
-        for name, macro in d[source_module].items():
-            seen_names.add(name)
-            if assignments == "ALL":
-                d[target_module][mangle(prefix + name)] = macro
-            elif name in assignments:
-                d[target_module][mangle(prefix + assignments[name])] = macro
+    assignments: str or list of tuples of strs
+        The string "ALL" or a dict of macro names to aliases.
 
-    if assignments != "ALL":
-        unseen = frozenset(assignments.keys()).difference(seen_names)
-        if unseen:
-            raise ImportError("cannot require names: " + repr(list(unseen)))
+    prefix: str, optional ("")
+        If nonempty, its value is prepended to the name of each imported macro.
+        This allows one to emulate namespaced macros, like
+        "mymacromodule.mymacro", which looks like an attribute of a module.
 
+    Returns
+    -------
+    out: None
+    """
+    if not inspect.ismodule(source_module):
+        source_module = importlib.import_module(source_module)
+
+    source_macros = source_module.__dict__.setdefault('__macros__', {})
+    source_tags = source_module.__dict__.setdefault('__tags__', {})
+
+    if len(source_module.__macros__) + len(source_module.__tags__) == 0:
```

Comment:
I think you could write this as `not source_module.__macros__  and not source_module.__tags__`. But actually, I don't see the point of this check. Shouldn't requiring a module with no macros just not pull in any macros? Or would that have some unexpected consequence?

Reply:
Perhaps it's reasonable to silently allow a `(require module)` when `module` has no macros or tags, but I'm on the fence about that.  Regardless, that check short-circuits the steps that follow and provides a much more precise error when `assignments` are not just `ALL`.